### DISCUSSION TITLE
how can i refer to the class name in my model field?

### DISCUSSION THREAD

**Discussion by cfal** [pydantic/pydantic#3906] [Question] [answered]
I have a field that i'd like to be populated with the class name. Unfortunately, the only way I've found to do it right now is by overriding `dict()`, which means that it's not correctly set before serialization:

```
class Creature(BaseModel):
  creature_type: str
  def dict(self, *args, **kwargs):
    result = super().dict(*args, **kwargs)
    result["creature_type"] = self.__class__.__name__
    return result

# serializing this would then have {"creature_type": "Cat", "can_meow": True}
class Cat(Creature):
  can_meow: bool = True
```

I've considered the following:
- using `__init__` to set `self.creature_type` - but `__init__` isn't always called (not sure if relevant, but I'm using pydantic via FastAPI)
- using a `default_factory` - but default_factory doesn't get passed any arguments and I can't refer back to the class name.

Is there a better way to do this?

Thank you!

**Comment by PrettyWood** [comment, marked as answer]
Hi @cfal 
Probably easier to use root_validator like this
```py
from pydantic import BaseModel, root_validator


class Creature(BaseModel):
  creature_type: str

  @root_validator(pre=True)
  def set_creature_type(cls, values):
      values['creature_type'] = cls.__name__
      return values
```
Hope it helps

**Comment by cfal** [comment]
awesome, thank you @PrettyWood! validators was just what i needed - ended up with this:

```
@validator(pre=True, always=True)
def set_type(cls, v):
  return v or cls.__name__
```

**Reply by rjurney** [reply]
Thank you @cfal this is SO helpful. I looked far and wide for an hour before I found this!

**Reply by bra-fsn** [reply]
BTW, both `validator` and `root_validator` are deprecated in pydantic 2
